update(deluge-web-client): typing and documentation

zakkarry opened the pull request:
It seems mostly good, the API calls all looked good. I did make some changes to the documentation on certain functions that were either missing params and their usages or incorrectly labeled (possibly due to a copy paste)

One of the main things I noticed was you were using id 1 for everything, and if the class is instanciated rather than used sporadically, re-using ID 1 is not ideal. The connection should increment the ID as it sends queries, otherwise things can get messy. I put a variable on self called `ID` and set it to 1 in the constructor, then added `self.ID += 1` in execute so it is incremented.

There was a type issue on `__exit__` returning NoneType when you return False - so I corrected that typing.

Lastly I changed your positional arguments that did not match up to non-positional so it would pass correctly.

I'm not sure if you received any syntax errors or type errors during testing, but it improves readability and maintainability or anyone outside of you for them to be explicit.

I did not test any of these changes, by the way, I have company and just wanted to run through this and point out some things. Sorry about that, but I assume based on what I see that you have a pretty good grasp.

I also moved requirements.txt to root, as at leat in my IDE and normal practice is to have it in the root - my IDE wont even pick up that there is a requirements.txt where you had it (in docs). Project files should reflect the new location as well.

I like your work, very glad to have someone like you in the community making a contribution like this :)

Take what you like, toss away what you don't - this is just some thoughts for you :)

Also if you want to automate publishing to pypi with publishing on github via a workflow for GHA let me know.

zakkarry committed:
allow NoneType for Response

NoneType is used in the class without being defined with that type

zakkarry committed:
move requirements.txt to root

zakkarry committed:
update docs - use incremental id for requests - non-positional args

jessielw commented on deluge_web_client/client.py:
```diff
@@ -78,26 +79,14 @@ def _connect_to_first_host(self, timeout: int) -> dict[str, Response]:
             )
         }
 
-    def close_session(self) -> None:
-        """
-        Closes the `DelugeWebClient` session.
-
-        This is handled automatically
-        when `DelugeWebClient` is used in a context manager.
-        """
-        self.session.close()
-
```
Should we still provide a way for the user to close the session? I noticed you deleted this but didn't see why in your comment.

zakkarry replied:
that wasnt intentional, you pushed 1.0.2 while i was editing and i copied the files back over, def should readd. sorry about that, shoulda rebased instead

jessielw replied:
Ahhh, that makes sense!

jessielw commented on deluge_web_client/client.py:
```diff
@@ -23,9 +25,8 @@ def __enter__(self) -> "DelugeWebClient":
-    def __exit__(self, exc_type, exc_value, traceback) -> None:
-        """End of with statement, closes the session."""
-        self.close_session()
+    def __exit__(self, exc_type, exc_value, traceback) -> bool:
+        """End of with statement."""
         return False
```
Thanks for this! We could actually remove the return statement and just `-> None` tbh. I don't think it was necessary to return `False`. 

Also if we re-add `close_session()` I think it makes sense to be put back here right?

zakkarry replied:
Yes, I don't see what the return value actually provides, and given that its always False, you could just `return` - yielding NoneType.

jessielw replied:
Sounds good, if you would like to make that change as well!

jessielw commented on requirements.txt:
```diff
@@ -0,0 +1,3 @@
+sphinx==7.4.7
+sphinx-autodoc-typehints==2.3.0
+furo==2024.8.6
```
This was actually a `readthedocs` requirement.txt file. It should stay within the docs folder. The requirements are handled via `poetry` instead of the older requirements.txt method.

zakkarry replied:
Oh my mistake. I'd suggest adding a requirements.txt for the project - it was late and I didn't investigate that all too much and just put it in root. I'll fix this though and generate a requirements.txt for the actual project if you'd like.

jessielw replied:
Thanks! I appreciate it, once you make those changes I can merge! 

I never really thought about having the `project.toml` and `requirements.txt`. I did a quick search and it is recommended to include both. So sure you can add that in as well if you want!

zakkarry replied:
Ok, I moved the docs requirements.txt back and put a requirements.txt with just requests (that's it?) in root. Sorry about that.

I also formatted the main .py with black - it changed some indenting depth, but otherwise mostly the same.

Readded the exit to session close, (once again sorry about being out of date and not rebasing)

I think I covered everything?

jessielw replied:
Look's good!

jessielw commented:
> It seems mostly good, the API calls all looked good. I did make some changes to the documentation on certain functions that were either missing params and their usages or incorrectly labeled (possibly due to a copy paste)
> 
> One of the main things I noticed was you were using id 1 for everything, and if the class is instanciated rather than used sporadically, re-using ID 1 is not ideal. The connection should increment the ID as it sends queries, otherwise things can get messy. I put a variable on self called `ID` and set it to 1 in the constructor, then added `self.ID += 1` in execute so it is incremented.
> 
> There was a type issue on `__exit__` returning NoneType when you return False - so I corrected that typing.
> 
> Lastly I changed your positional arguments that did not match up to non-positional so it would pass correctly.
> 
> I'm not sure if you received any syntax errors or type errors during testing, but it improves readability and maintainability or anyone outside of you for them to be explicit.
> 
> I did not test any of these changes, by the way, I have company and just wanted to run through this and point out some things. Sorry about that, but I assume based on what I see that you have a pretty good grasp.
> 
> I also moved requirements.txt to root, as at leat in my IDE and normal practice is to have it in the root - my IDE wont even pick up that there is a requirements.txt where you had it (in docs). Project files should reflect the new location as well.
> 
> I like your work, very glad to have someone like you in the community making a contribution like this :)
> 
> Take what you like, toss away what you don't - this is just some thoughts for you :)
> 
> Also if you want to automate publishing to pypi with publishing on github via a workflow for GHA let me know.

Thanks for taking the time to go over this. I have a few more questions as well. 

Do you use `mypy` for type hint checking? ATM I don't use a type hint checker at all, so it's likely something I should incorporate. I use **vscode** with **PyLance**. There is no errors as far as I can see with typing with this setup. 
Edit: I see that type checking was set to **off** by default, I guess for performance reasons. 

 When it comes to these two methods https://github.com/jessielw/deluge-web-client/blob/afd4607c57fbabe456d546eea3699bf7b236d5ba/deluge_web_client/client.py#L356-L407, I didn't really understand how the keys/diff worked. I implemented them based on the source code, but I wasn't able to test things out. How exactly does those work? 
 
 Thanks for taking the time again! I appreciate it!

zakkarry commented:
I am actually not familiar with what the diff key does tbh, ive never included it for any calls ive made, but i can look at the codebase and see what it does, although i suspect it's probably a very niche use case or something.

if i cant immediately tell, ill ask another team member, but itll be in a little, im mobile right now.

generally speaking, web.update_ui was what i used for listing and single get torrent status for details past the keys available in update ui.

ill look into it though and get back to you.

as for ide i use pycharm but type hinting should be on if youre using types, my ide immediately pointed the bad typing out and i mainly code it typescript, so it kinda screams to me to fix it.

overall it looked pretty good man. i can take a more granular look later and actually test things in a dev environment to iron out what the ux is like.

jessielw commented:
> I am actually not familiar with what the diff key does tbh, ive never included it for any calls ive made, but i can look at the codebase and see what it does, although i suspect it's probably a very niche use case or something.
> 
> if i cant immediately tell, ill ask another team member, but itll be in a little, im mobile right now.

No big deal if you can't. As long as it's included to be passed I think that's good enough for the users that may use it potentially. I don't think I personally have any use case for it right now. 

> as for ide i use pycharm but type hinting should be on if youre using types, my ide immediately pointed the bad typing out and i mainly code it typescript, so it kinda screams to me to fix it.

I used to use PyCharm at the start. I moved to VsCode when I started coding in numerous other languages over time. PyLint is built in, but I'm going to search around to get more information on what would be expected, so it could be ran as well as needed automatically.

Edit: It looks like `MyPy` might be the go to. 

> overall it looked pretty good man. i can take a more granular look later and actually test things in a dev environment to iron out what the ux is like.

zakkarry commented:
> > I am actually not familiar with what the diff key does tbh, ive never included it for any calls ive made, but i can look at the codebase and see what it does, although i suspect it's probably a very niche use case or something.
> > if i cant immediately tell, ill ask another team member, but itll be in a little, im mobile right now.
> 
> No big deal if you can't. As long as it's included to be passed I think that's good enough for the users that may use it potentially. I don't think I personally have any use case for it right now.
> 
> > as for ide i use pycharm but type hinting should be on if youre using types, my ide immediately pointed the bad typing out and i mainly code it typescript, so it kinda screams to me to fix it.
> 
> I used to use PyCharm at the start. I moved to VsCode when I started coding in numerous other languages over time. PyLint is built in, but I'm going to search around to get more information on what would be expected, so it could be ran as well as needed automatically.
> 
> Edit: It looks like `MyPy` might be the go to.
> 
> > overall it looked pretty good man. i can take a more granular look later and actually test things in a dev environment to iron out what the ux is like.

Yea, I have a jetbrains ultimate license for another OSS project for free, I like their IDEs alot, intelliJ and pycharm are great for me.

I commented on your review. Not sure if I missed anything, just let me know.

I don't assume the type additions I added will cause any issues, but if you want to test things before merging or requesting additional changes to that end let me know.

zakkarry commented:
How are you publishing to PyPi right now? I don't see any workflows?

jessielw commented:
> How are you publishing to PyPi right now? I don't see any workflows?

Currently just manually utilizing poetry. 

```
poetry build
poetry publish
```

But I would like to automate it. I haven't looked into how to do this yet with github.

zakkarry commented:
> > How are you publishing to PyPi right now? I don't see any workflows?
> 
> Currently just manually utilizing poetry.
> 
> ```
> poetry build
> poetry publish
> ```
> 
> But I would like to automate it. I haven't looked into how to do this yet with github.

It's actually pretty easy, you can just go to Actions and hit new workflow, it should be pretty much the default - and then add your PyPi api token for either your account or the project to the repo's secrets.

Here's my projects pypi if you want to reference, iirc its literally just "PyPi publishing" default.

https://github.com/zakkarry/retraktarr/blob/master/.github/workflows/pypi-publish.yml

zakkarry commented:
Whenever you publish a release through github, it will push it to pypi.
